statsd/receiver.go

```diff
@@ -5,133 +5,116 @@ import (
+	tags      types.Tags              // Tags to add to all metrics
 }
 
-// ListenAndReceive listens on the UDP network address of srv.Addr and then calls
-// Receive to handle the incoming datagrams. If Addr is blank then DefaultMetricsAddr is used.
-func (mr *MetricReceiver) ListenAndReceive() error {
-	addr := mr.Addr
-	if addr == "" {
-		addr = defaultMetricsAddr
-	}
-	c, err := net.ListenPacket("udp", addr)
-	if err != nil {
-		return err
-	}
-
-	mq := make(messageQueue, maxQueueSize)
-	for i := 0; i < mr.MaxMessengers; i++ {
-		go mq.dequeue(mr)
+// NewMetricReceiver initialises a new Receiver.
+func NewMetricReceiver(ns string, tags []string, cloud cloudprovider.Interface, handler Handler) Receiver {
+	return &metricReceiver{
+		cloud:     cloud,
+		handler:   handler,
+		namespace: ns,
+		tags:      tags,
 	}
-	for i := 0; i < mr.MaxReaders; i++ {
-		go mr.receive(c, mq)
-	}
-	return nil
 }
 
-// increment allows counting server stats using default tags.
-func (mr *MetricReceiver) increment(name string, value int) {
-	mr.Handler.HandleMetric(types.NewMetric(internalStatName(name), float64(value), types.COUNTER, mr.Tags))
-}
-
-type messageQueue chan message
-
-func (mq messageQueue) enqueue(m message, mr *MetricReceiver) {
-	mq <- m
-}
-
-func (mq messageQueue) dequeue(mr *MetricReceiver) {
-	for m := range mq {
-		mr.handleMessage(m.addr, m.msg)
-		runtime.Gosched()
+// GetStats returns current Receiver stats. Safe for concurrent use.
+func (mr *metricReceiver) GetStats() ReceiverStats {
+	return ReceiverStats{
+		time.Unix(0, atomic.LoadInt64(&mr.lastPacket)),
+		atomic.LoadUint64(&mr.badLines),
+		atomic.LoadUint64(&mr.packetsReceived),
+		atomic.LoadUint64(&mr.metricsReceived),
 	}
 }
 
-// receive accepts incoming datagrams on c and calls mr.handleMessage() for each message.
-func (mr *MetricReceiver) receive(c net.PacketConn, mq messageQueue) {
-	defer c.Close()
-
-	var buf []byte
+// Receive accepts incoming datagrams on c, parses them and calls Handler.HandleMetric() for each metric.
+func (mr *metricReceiver) Receive(ctx context.Context, c net.PacketConn) error {
+	buf := make([]byte, packetSizeUDP)
 	for {
-		if len(buf) < packetSizeUDP {
-			buf = make([]byte, packetBufSize, packetBufSize)
-		}
-
+		// This will error out when the socket is closed.
 		nbytes, addr, err := c.ReadFrom(buf)
 		if err != nil {
-			log.Printf("Error %s", err)
+			if netErr, ok := err.(net.Error); ok && !netErr.Temporary() {
+				select {
+				case <-ctx.Done():
+				default:
+					return fmt.Errorf("non-temporary error reading from socket: %v", err)
+				}
+				return nil
+			}
+			log.Warnf("Error reading from socket: %v", err)
 			continue
 		}
-		msg := buf[:nbytes]
-		mr.increment("packets_received", 1)
-		mq.enqueue(message{addr, msg}, mr)
-		buf = buf[nbytes:]
-		runtime.Gosched()
+		// TODO consider updating counter for every N-th iteration to reduce contention
+		atomic.AddUint64(&mr.packetsReceived, 1)
```

Comment:
Atomic counter is much faster than creating a metric object and processing it (hopefully just a [lock xadd](http://x86.renejeschke.de/html/file_module_x86_id_327.html) instruction if compiler authors did a good job :) ). Current implementation effectively processes ~2 additional metrics per metric received - 66% overhead! (`packets_received` and `metrics_received` for each packet). This new implementation collects stats from aggregators and receiver and sends them in `Flusher`.